Sources/TuistKit/Commands/TaskCommand.swift

```diff
@@ -0,0 +1,183 @@
+            return "You must provide task name"
+        }
+    }
+
+    var type: ErrorType {
+        switch self {
+        case .taskNotProvided:
+            return .bug
+        }
+    }
+}
+
+struct TaskCommand: ParsableCommand {
+    static var configuration: CommandConfiguration {
+        CommandConfiguration(
+            commandName: "task",
+            abstract: "Runs task defined in Tasks.swift"
+        )
+    }
+
+    @Argument(
+        help: "Name of a task you want to run"
+    )
+    var task: String
+
+    @Option(
+        name: .shortAndLong,
+        help: "The path to the directory where the tasks are run from",
+        completion: .directory
+    )
+    var path: String?
+
+    func run() throws {
+        try TaskService().run(
+            task,
+            options: taskOptions,
+            path: path
+        )
+    }
+
+    init() {}
+
+    var taskOptions: [String: String] = [:]
+
+    // Custom decoding to decode dynamic options
+    init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        task = try container.decode(Argument<String>.self, forKey: .task).wrappedValue
+        path = try container.decodeIfPresent(Option<String>.self, forKey: .path)?.wrappedValue
+        try TaskCommand.requiredOptions.forEach { option in
+            taskOptions[option.name] = try container.decode(
+                Option<String>.self,
+                forKey: .required(option.name)
+            ).wrappedValue
+        }
+        try TaskCommand.optionalOptions.forEach { option in
+            guard let value = try container.decode(
+                Option<String?>.self,
+                forKey: .optional(option.name)
+            )
+            .wrappedValue
+            else { return }
+            taskOptions[option.name] = value
+        }
+    }
+}
+
+// MARK: - Preprocessing
+
+extension TaskCommand {
+    static var requiredOptions: [(name: String, option: Option<String>)] = []
+    static var optionalOptions: [(name: String, option: Option<String?>)] = []
+
+    /// We do not know template's option in advance -> we need to dynamically add them
+    static func preprocess(_ arguments: [String]? = nil) throws {
+        guard
+            let arguments = arguments,
+            arguments.count >= 2
+        else { throw TaskCommandError.taskNotProvided }
+        guard !configuration.subcommands.contains(where: { $0.configuration.commandName == arguments[1] }) else { return }
+        // We want to parse only the name of a task, not its arguments which will be dynamically added
+        // Plucking out path arguments
+        let pairedArguments: [[String]] = stride(from: 2, to: arguments.count, by: 2).map {
+            Array(arguments[$0 ..< min($0 + 2, arguments.count)])
+        }
+        let filteredArguments = pairedArguments
+            .filter {
+                $0.first == "--path" || $0.first == "-p"
+            }
+            .flatMap { $0 }
```

Comment:
Could we share some of this with the template command or are they distinct

Reply:
I'd try to keep the current code for these dynamic commands "akward", so we don't overuse them and use only when there is no other way - also as a way to force myself to try to push some dynamism to the argument parser itself.